src/chatCompletionEventDataFactory.ts

```diff
@@ -0,0 +1,141 @@
+import { v4 as uuid } from 'uuid';
+import {
+  Configuration,
+  CreateChatCompletionRequest,
+  CreateChatCompletionResponse,
+} from 'openai';
+
+import {
+  ChatCompletionSummaryAttributes,
+  EventAttributes,
+  EventData,
+} from './eventTypes';
+import {
+  filterUndefinedValues,
+  isString,
+  removeUndefinedValues,
+} from './utility';
+import { EventAttributesBuilder } from './eventAttributesBuilder';
+
+export type RequestHeaders = Record<
+  string,
+  string | number | boolean | undefined
+>;
+
+export interface ChatCompletionEventDataFactoryOptions {
+  request: CreateChatCompletionRequest;
+  responseData: CreateChatCompletionResponse;
+  responseTime: number;
+  headers?: RequestHeaders;
+  openAiConfiguration?: Configuration;
+}
+
+export const createChatCompletionEventDataFactory = () => {
+  const createEventDataList = (
+    options: ChatCompletionEventDataFactoryOptions,
+  ): EventData[] => {
+    const completionId = uuid();
+
+    const messageDataList = createMessageEventDataList(completionId, options);
+    const summaryData = createSummaryEventData(completionId, options);
+
+    return [...messageDataList, summaryData];
+  };
+
+  const createMessageEventDataList = (
+    completion_id: string,
+    { request, responseData }: ChatCompletionEventDataFactoryOptions,
+  ): EventData[] => {
+    return getMessages(request, responseData).map<EventData>(
+      (message, sequence) => ({
+        eventType: 'LlmChatCompletionMessage',
+        attributes: {
+          id: uuid(),
+          sequence,
+          completion_id,
+          content: message.content,
+          role: message.role,
+          model: request.model,
+          vendor: 'openAI',
+        },
+      }),
+    );
+  };
+
+  const createSummaryEventData = (
+    id: string,
+    {
+      request,
+      responseData,
+      responseTime: response_time,
+      headers,
+      openAiConfiguration,
+    }: ChatCompletionEventDataFactoryOptions,
+  ): EventData => {
+    const { usage, choices } = responseData;
+
+    const initialAttributes: ChatCompletionSummaryAttributes = {
+      id,
+      response_time,
+      timestamp: Date.now(),
+      number_of_messages: getMessages(request, responseData).length,
+      vendor: 'openAI',
+      finish_reason: choices[choices.length - 1].finish_reason,
+      prompt_tokens: usage?.prompt_tokens,
```

Comment:
we should add "applicationName" to the attributes

Reply:
@nhoffmann-newrelic we get these fields implicitly.

Reply:
@vsiomin-newrelic @amir-newrelic Shouldn't "applicationName" be "application_name"?

Reply:
@nhoffmann-newrelic In the entity definition it's defined as applicationName  https://github.com/newrelic/entity-definitions/blob/4c4b314425934c3af6fcb29c36998a14181f3367/definitions/mlops-llm_application/definition.yml#LL5C1-L6C1